File: android/entityparser/EntityScore.java
package entityparser;

import java.util.Dictionary;
import java.util.Hashtable;

/**
 * EntityScore
 *
 * Handles scoring of ngram entities using generated/loaded dictionary vocabulary
 */
public class EntityScore {
    // TODO: correct load
    private static Dictionary ngramVocab = new Hashtable();

    /**
     * offline scoring mechanism, references dicitonary of keys
     * @param words word to check if exists
     * @return score of word, (1 for unigram, > 1 for identified ngrams, < 1 for unidentified ngrams)
     */
    public static double getNgramScore(String words) {
        double recognizedNgramOffset = 0.10;
        double unknownNgramOffset = 0.15;
        int wordsCount = words.split("\\s").length;
        double score = 1;

        if (wordsCount > 1) {
            if (ngramVocab.get(words) != null) {
                score += (recognizedNgramOffset * wordsCount);
            }
            else {
                score += (unknownNgramOffset * wordsCount);
            }
        }
        return score;
    }
}

File: online_api/api/entity_phrase_parser/entity_score.py
"""
@author: Richard Valente <rcvalente@wpi.edu>
@date: 1/4/2020
@contributors: rcvalenteai

EntityScore

Handles scoring of ngram entities using generated/loaded dictionary vocabulary
"""
from api.entity_phrase_parser.vocab_generator import *


def get_ngram_score(word):
    """
    offline scoring mechanism for prioritizing larger compound words, references dictionary of keys
    :param word: word to check if exists
    :return: score of word, (1 for unigram, > 1 for identified ngrams, < 1 for unidentified ngrams)
    """
    recognized_ngram_offset = 0.10
    unknown_ngram_offset = 0.15
    words_count = len(word.split())
    score = 1
    if words_count > 1:
        if word in ngram_vocab.keys():
            score += (recognized_ngram_offset * words_count)
        else:
            score -= (unknown_ngram_offset * words_count)
    return score


ngram_vocab = initialize_vocabulary()
